### ISSUE TITLE
Batch query bytes error

### ISSUE THREAD

**Issue by chzhuo** [aerospike/aerospike-server#86]
#Environment
Server : Aerospike Community Edition build 3.5.13
Client : 
<dependency>
	<groupId>com.aerospike</groupId>
	<artifactId>aerospike-client</artifactId>
	<version>3.1.4</version>
</dependency> 

Configure:  as/etc/aerospike_dev.conf

I use 'make init  ;  make start'  to start the server.

#Run Code
```java
	public static void main(String[] args) {
		AerospikeClient client = new AerospikeClient("192.168.1.134", 3000);
		int size = 3;
		//insert
		byte[] bytes = new byte[1024*512];
		for(int i = 0;i<size;++i){
			Key key = new Key("test","demo",i);
			Bin bin = new Bin("value",bytes);
			client.put(null, key, bin);
		}
		
		//query 
		Key[] keys = new Key[size];
		for(int i = 0;i<size;++i){
			keys[i] = new Key("test","demo",i);
		}
		Record[] records = client.get(null, keys);
		System.out.println(records.length);
	}
```
I catch the exception
```java
Exception in thread "main" com.aerospike.client.AerospikeException$Parse: Error Code -2: Unexpected batch key returned: test,ae6a515f37de4adacc4496e2da1fea848c3a9270,1,0
	at com.aerospike.client.command.Batch$GetArrayDirect.parseRow(Batch.java:172)
	at com.aerospike.client.command.MultiCommand.parseGroup(MultiCommand.java:114)
	at com.aerospike.client.command.MultiCommand.parseResult(MultiCommand.java:71)
	at com.aerospike.client.command.SyncCommand.execute(SyncCommand.java:56)
	at com.aerospike.client.command.BatchExecutor.execute(BatchExecutor.java:65)
	at com.aerospike.client.AerospikeClient.get(AerospikeClient.java:630)
	at com.example.aerospike_java.local.QQ.main(QQ.java:48)
```

**Comment by billbargens** [comment]
I met the same problem.  

Even worse, sometimes the server will crashed down with the fellow message:

Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::160) SIGSEGV received, aborting Aerospike Community Edition build 3.5.15
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: found 11 frames
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 0: /disk1/aerospike/cluster/bin/asd(as_sig_handle_segv+0x59) [0x454414]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 1: /lib64/libc.so.6() [0x34bca326a0]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 2: /disk1/aerospike/cluster/bin/asd(free+0x57) [0x56e017]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 3: /disk1/aerospike/cluster/bin/asd(cf_free_at+0x7e) [0x4e3e4e]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 4: /disk1/aerospike/cluster/bin/asd(cf_buf_builder_reserve_internal+0xde) [0x4e5b22]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 5: /disk1/aerospike/cluster/bin/asd(cf_buf_builder_reserve+0x4d) [0x4e5b9a]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 6: /disk1/aerospike/cluster/bin/asd(as_msg_make_response_bufbuilder+0x439) [0x44dac6]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 7: /disk1/aerospike/cluster/bin/asd() [0x467eff]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 8: /disk1/aerospike/cluster/bin/asd(batch_process_queue+0x85) [0x4680ea]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 9: /lib64/libpthread.so.0() [0x34bce079d1]
Aug 20 2015 15:48:40 GMT: WARNING (as): (signal.c::162) stacktrace: frame 10: /lib64/libc.so.6(clone+0x6d) [0x34bcae89dd]

**Comment by billbargens** [comment]
When I rolled back the server to version 3.5.8,  the problem above disappeared!

**Comment by MaxJust** [comment]
have same problem

**Comment by bbulkow** [comment]
Wasn't this listed as resolved at least with 3.5.15 from July 21 ( 5 weeks ago ) ?
The 3.5.13 release is no longer on Aerospike's website, I believe because of this issue.

**Comment by chzhuo** [comment]
Have fixed in 3.6.0

**Comment by klauss42** [comment]
Just got a similar crash (same error message) with release 3.6.4

**Comment by wchu-citrusleaf** [comment]
Can you post your stack? and number of nodes? client used? We might also ask for more detail later.
Thanks.

**Comment by klauss42** [comment]
I use the Java client version 3.1.6, servers just upgraded to 3.6.4, cluster of 6 servers. Here ist the Java stacktrace:
```
java.util.concurrent.ExecutionException: com.aerospike.client.AerospikeException$Parse: Error Code -2: Unexpected batch key returned: null,,0
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:1.7.0_67]
	at java.util.concurrent.FutureTask.get(FutureTask.java:188) ~[na:1.7.0_67]
<snip>
	at com.aerospike.client.async.AsyncQueryExecutor.onSuccess(AsyncQueryExecutor.java:55) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.AsyncMultiExecutor.childSuccess(AsyncMultiExecutor.java:55) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.AsyncMultiCommand.onSuccess(AsyncMultiCommand.java:227) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.AsyncCommand.finish(AsyncCommand.java:293) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.AsyncMultiCommand.read(AsyncMultiCommand.java:101) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.SelectorManager.processKey(SelectorManager.java:165) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.SelectorManager.runCommands(SelectorManager.java:107) [aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.async.SelectorManager.run(SelectorManager.java:68) [aerospike-client-3.1.6.jar:na]
Caused by: com.aerospike.client.AerospikeException$Parse: Error Code -2: Unexpected batch key returned: null,,0
	at com.aerospike.client.command.Batch$GetArrayCommand.parseRow(Batch.java:120) ~[aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.command.MultiCommand.parseGroup(MultiCommand.java:114) ~[aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.command.MultiCommand.parseResult(MultiCommand.java:71) ~[aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.command.SyncCommand.execute(SyncCommand.java:56) ~[aerospike-client-3.1.6.jar:na]
	at com.aerospike.client.command.Executor$ExecutorThread.run(Executor.java:127) ~[aerospike-client-3.1.6.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) ~[na:1.7.0_67]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) ~[na:1.7.0_67]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.7.0_67]
```

In the server log the last lines before the crash are the following:
```
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x74aeb62c7dfd694318bcbba057829b6fa8dca73a
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x9eaca6971cd0466f0ef644e72ab750e8c5069856
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0xb80884a20912396d17c1caf55b0e8299bf00438f
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0xc46fb2b49924c3cdfbb90185ebbde829abd87194
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x51e972699d790527e5422292e565c3037fc5fa28
Nov 13 2015 16:15:47 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x05603ef953803dffbadcd30b2bd77cb0783d8f03
Nov 13 2015 16:15:48 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x628822bc241ab53b09799b7116b8276e241f6d5e
Nov 13 2015 16:15:48 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x0eced2ba559918cc724d7ed24cfdaf2251c27ef5
Nov 13 2015 16:15:48 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x19d8d4a8663d13cdb6810f511de3a89c414c6b0d
Nov 13 2015 16:15:48 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0xde4dcc5a9662c4af81f9b42df53756835b492723
Nov 13 2015 16:15:50 GMT: INFO (drv_ssd): (drv_ssd.c::2088) device /opt/aerospike/data/loadtest: used 16177236608, contig-free 9851M (9851 wblocks), swb-free 1, w-q 0 w-tot 985 (0.1/s), defrag-q 0 defrag-tot 769 (0.0/s) defrag-w-tot 53 (0.0/s)
Nov 13 2015 16:15:51 GMT: WARNING (proxy): (thr_proxy.c::381) SHIPPED_OP ORIG Missing fd in proto_fd <Digest>:0x3791375886949963e480924e30e31c30773078b1
Nov 13 2015 16:15:52 GMT: WARNING (as): (signal.c::161) SIGSEGV received, aborting Aerospike Community Edition build 3.6.4 os ubuntu12.04
Nov 13 2015 16:15:52 GMT: WARNING (as): (signal.c::163) stacktrace: found 2 frames
Nov 13 2015 16:15:52 GMT: WARNING (as): (signal.c::163) stacktrace: frame 0: /usr/bin/asd(as_sig_handle_segv+0x5d) [0x473836]
Nov 13 2015 16:15:52 GMT: WARNING (as): (signal.c::163) stacktrace: frame 1: /lib/x86_64-linux-gnu/libc.so.6(+0x36d40) [0x7f027afefd40]
```

**Comment by klauss42** [comment]
just crashed another server in the cluster. The last lines in aerospike.log are: 
````
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::161) SIGSEGV received, aborting Aerospike Community Edition build 3.6.4 os ubuntu12.04
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: found 10 frames
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 0: /usr/bin/asd(as_sig_handle_segv+0x5d) [0x473836]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 1: /lib/x86_64-linux-gnu/libc.so.6(+0x36d40) [0x7efff03e0d40]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 2: /usr/bin/asd() [0x58a142]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 3: /usr/bin/asd() [0x5a1a85]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 4: /usr/bin/asd(free+0x350) [0x57de00]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 5: /usr/bin/asd(cf_free_at+0x7e) [0x4f8458]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 6: /usr/bin/asd(cf_rchash_reduce_v+0xb3) [0x5258ba]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 7: /usr/bin/asd(as_smd_thr+0x1453) [0x485095]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 8: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8182) [0x7efff12b8182]
Nov 13 2015 16:31:21 GMT: WARNING (as): (signal.c::163) stacktrace: frame 9: /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7efff04a447d]
````

**Comment by BrianNichols** [comment]
Are you requesting records that have LDT bins in your batch call?

**Comment by klauss42** [comment]
Yes, there is 1 largeset  bin

Klaus

**Comment by MaxJust** [comment]
I have a problem with LDS too. I wrote about it a few weks ago.

**Comment by BrianNichols** [comment]
It looks like there is some instability when requesting LDT bins which can't be returned by batch.  Can you try using batch bin name filters that include all bins you need except the LDT bin?

For batch where all bins are same:
```java
Record[] records = client.get(null, keys, binName1, binName2);
```
For batch where bins are different for each record:
```java
List<BatchRead> records = new ArrayList<BatchRead>();
String[] binFilter1 = new String[] {bin1, bin2};
records.add(new BatchRead(new Key(ns, set, key1), binFilter1));
records.add(new BatchRead(new Key(ns, set, key2), binFilter1));
String[] binFilter2 = new String[] {bin3, bin4, bin5};
records.add(new BatchRead(new Key(ns, set, key3), binFilter2));

client.get(null, records);
```

**Comment by xiaosuo** [comment]
wow，it seems 3.6.4 still have this issue. The good news is I don't need to verify it.

**Comment by klauss42** [comment]
Update: I downgraded servers to 3.6.2 again. Reran my test while migrations were still in progress and experienced a crash of one server again. After migrations went through, the test runs successful again. The test is a iteration over all my records in 2 tables (total ~64 million records). So maybe this has something to do with migrations in progress?
In my test with 3.6.4 I was impatient so I ran the test while migrations were in progress.

Maybe this helps analyzing the issue
